homeassistant/components/dht/sensor.py

```diff
@@ -160,18 +171,27 @@ def update(self):
 class DHTClient:
     """Get the latest data from the DHT sensor."""
 
-    def __init__(self, adafruit_dht, sensor, pin):
+    def __init__(self, sensor, pin, name):
         """Initialize the sensor."""
-        self.adafruit_dht = adafruit_dht
         self.sensor = sensor
-        self.pin = pin
+        self.pin = getattr(board, pin)
         self.data = {}
+        self.name = name
 
     @Throttle(MIN_TIME_BETWEEN_UPDATES)
     def update(self):
         """Get the latest data the DHT sensor."""
-        humidity, temperature = self.adafruit_dht.read_retry(self.sensor, self.pin)
-        if temperature:
-            self.data[SENSOR_TEMPERATURE] = temperature
-        if humidity:
-            self.data[SENSOR_HUMIDITY] = humidity
+        try:
+            dht = self.sensor(self.pin)
+            humidity = dht.humidity
+            temperature = dht.temperature
+            if temperature:
+                self.data[SENSOR_TEMPERATURE] = temperature
+            if humidity:
+                self.data[SENSOR_HUMIDITY] = humidity
+        except RuntimeError:
+            _LOGGER.debug("Unexpected value from DHT sensor: %s", self.name)
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Error updating DHT sensor: %s", self.name)
+        finally:
+            dht.exit()
```

Comment:
This will sweep the library bugs under the rug, while not unnecessarily eating errors that might be important for debugging legitimate problems. Untested.

```suggestion
        dht = self.sensor(self.pin)
        try:
            attempts = 0
            while True:
                try:
                    # .humidity and .temperature are properties which trigger a
                    # read of the sensor if it has been more than 2 seconds
                    # since the last read, or return the last value otherwise.
                    #
                    # The library is very buggy, and often fails to read all
                    # the pulses or accurately measure their timing. The issue
                    # is worse when the CPU is under load. When it doesn't get
                    # the right number of pulses or the checksum fails it
                    # throws a RuntimeError.
                    #
                    # This issue could be avoided by using pigpiod directly.
                    # See https://gist.github.com/bitglue/bf077095243405702367876af444b98a
                    # for example. That would have to be accepted upstream in
                    # the adafruit libs, or implemented and released as a
                    # separate library to be used here.
                    humidity = dht.humidity
                    temperature = dht.temperature
                except RuntimeError:
                    attempts += 1
                    if attempts >= 3:
                        raise
                    # The datasheet specifies a minimum of 2 seconds between
                    # readings. If we don't wait at least 2 seconds, the
                    # library will just return the last values rather than
                    # trigger a new reading.
                    time.sleep(2.1)
                else:
                    self.data[SENSOR_TEMPERATURE] = temperature
                    self.data[SENSOR_HUMIDITY] = humidity
                    break
        finally:
            dht.exit()
```

Reply:
We can't sleep in the thread pool since it risks exhausting the pool.

Reply:
Then we can't read the sensor either, because there's a sleep in the library.

The only reason to have threads is to run blocking code. If you didn't have blocking code then you could just run everything in the main loop.

This will block for at most 7 seconds. If there is a DHT22 that fails 100% of the time on each of the 28 GPIOs on a Raspberry Pi, and update is called at most once per 30 seconds, that's a mean concurrency of 7*28/30 = 6.5, and a worst case of 28.

Seems like a reasonable number of threads to me?